### ISSUE TITLE
BBIM Tool Join/Extend to be based on cursor location

### ISSUE THREAD

**Issue by myoualid** [IfcOpenShell/IfcOpenShell#3206]
Take this example where two walls are being joined (or trying to extend):
Is it possible to extend/ join based on the position of the Cursor Click when extending/joining? It seems as though the longer side of an object always wins ! 

https://github.com/IfcOpenShell/IfcOpenShell/assets/79010126/5f7a22e4-a72a-43a2-b85e-b64cc6b04162

**Comment by Gorgious56** [comment]
Yes I've wondered how to fix this issue, it's a bit cumbersome currently what I do is place the cursor somewhere along the bigger side, to make it smaller, again and again until I'm sure the other side is longer. And I've witnessed several times that after realizing I made a mistake and undoing with CTRL-Z the last "Extend", it somehow sends one of the wall limits flying several hundreds meters in the distance or plain removes the object from the file. Couldn't find a way to reliably replicate though, so it might be a PEBCAK ^^. 

I don't know if cursor click is manageable, the information may not be easily available using the API and there are other ways to select an object that don't involve clicking on the mesh. But maybe placing the 3D cursor on the side you want to extend / retract would be more straightforward ?

**Comment by myoualid** [comment]
Yeaaah, I also keep shortening the object to make sure it gets trimmed correctly but issa big pain :( My brain then starts to look for short elements to copy/paste before trimming. 

I suspect a popup would be just as time wasting. 

I hope we can crack this - imagine the collective time we can save for humanity 🥇

**Comment by Gorgious56** [comment]
So instead of an in-your-face popup, put the option to invert the side that gets trimmed in the redo panel that displays in the lower left of the viewport ? That way, if it goes according to plan, the user can continue their modelling, but if they don't like it, F9 > click on the toggle ?

**Comment by Moult** [comment]
Pro tip: use Cut instead of extend when you're unsure.

**Comment by theoryshaw** [comment]
@brunoperdigao could the raycasting part of your snapping code, be used in this fashion, such that when you click on an object, it knows 'where' on the object you clicked, and then can use that info to manage the order by which object trim/extend to each other?

**Comment by BIMvoice** [comment]
Built the redo-panel invert toggle you proposed, @Gorgious56, for both Join and Extend (#8902): an Invert Trim Side checkbox that shows up in the F9 Adjust Last Operation panel, so if the trim goes the wrong way you flip it there instead of shortening the wall until the guess flips. Default behaviour is unchanged.

@theoryshaw your raycasting-from-click idea is a natural follow-up: the Join and Extend gizmos already know their click position, and DumbWallJoiner.extend already resolves a similar ATSTART/ATEND decision from a 3D point, so it can be plumbed in as an optional bias point next. I also left a root-cause lead on this thread for the Ctrl-Z sends-a-wall-flying report (likely a desync between Blender's undo stack and IfcStore's own undo callback leaving the axis reference points mismatched).
